specs/phase1/validator.md

```diff
@@ -0,0 +1,529 @@
+
+```python
+def get_custody_bits_blocks() -> List[Bitlist[MAX_VALIDATORS_PER_COMMITTEE], MAX_SHARD_BLOCKS_PER_ATTESTATION]:
+    pass
+```
+
+##### Signature
+
+Set `attestation.signature = attestation_signature` where `attestation_signature` is obtained from:
+
+```python
+def get_attestation_signature(state: BeaconState,
+                              attestation: Attestation,
+                              privkey: int) -> BLSSignature:
+    domain = get_domain(state, DOMAIN_BEACON_ATTESTER, attestation.data.target.epoch)
+    attestation_data_root = hash_tree_root(attestation.data)
+    index_in_committee = attestation.aggregation_bits.index(True)
+    signatures = []
+    for block_index, custody_bits in enumerate(attestation.custody_bits_blocks):
+        custody_bit = custody_bits[index_in_committee]
+        signing_root = compute_signing_root(
+            AttestationCustodyBitWrapper(
+                attestation_data_root=attestation_data_root,
+                block_index=block_index,
+                bit=custody_bit,
+            ),
+            domain,
+        )
+        signatures.append(bls.Sign(privkey, signing_root))
+
+    return bls.Aggregate(signatures)
+```
+
+### Light client committee
+
+In addition to the core beacon chain responsibilities, Phase 1 adds an additional role -- the Light Client Committee -- to aid in light client functionality.
+
+Validators serve on the light client committee for `LIGHT_CLIENT_COMMITTEE_PERIOD` epochs and the assignment to be on a committee is known `LIGHT_CLIENT_COMMITTEE_PERIOD` epochs in advance.
+
+#### Preparation
+
+When `get_current_epoch(state) % LIGHT_CLIENT_COMMITTEE_PERIOD == LIGHT_CLIENT_COMMITTEE_PERIOD - LIGHT_CLIENT_PREPARATION_EPOCHS` each validator must check if they are in the next period light client committee by calling `is_in_next_light_client_committee()`.
+
+If the validator is in the next light client committee, they must join the `light_client_votes` pubsub topic to begin duties at the start of the next period.
+
+```python
+def is_in_next_light_client_committee(state: BeaconState, index: ValidatorIndex) -> boolean:
+    period_start_epoch = get_current_epoch(state) + LIGHT_CLIENT_COMMITTEE_PERIOD % get_current_epoch(state)
+    next_committee = get_light_client_committee(state, period_start_epoch)
+    return index in next_committee
+```
+
+#### Light clent vote
+
+During a period of epochs that the validator is a part of the light client committee (`validator_index in get_light_client_committee(state, epoch)`), the validator creates and broadcasts a `LightClientVote` at each slot.
+
+A validator should create and broadcast the `light_client_vote` to the `light_client_votes` pubsub topic when either (a) the validator has received a valid block from the expected block proposer for the current `slot` or (b) two-thirds of the `slot` have transpired (`SECONDS_PER_SLOT / 3` seconds after the start of `slot`) -- whichever comes _first_.
```

Comment:
Re timing, what do you think about 

- 0 second: beacon block proposer proposes a beacon block
- `SECONDS_PER_SLOT / 4` shard block proposer proposes a shard block
- `SECONDS_PER_SLOT 1 / 2`: attesters (including light client committee) build attestations
- `SECONDS_PER_SLOT 3 / 4` : aggregators (including light client) aggregate attestations

Reply:
yeah, I see the value in mirroring this in the roles

The main reason I staggered is because there are fewer steps in the LC flow so (1) better chance of full propagation between each step and (2) staggers local bursts of bandwidth and cpu between attestations and light client messages

thoughts on the above?

Reply:
Note to self. When phase1 run time is up, I'll provide the new numbers on proposer, attester, aggregator and LC's time spent for each operation

Reply:
@djrtwo 
I see. Sounds reasonable to stagger the traffic now and see if we can optimize it by the data from @terencechain :)